### PR TITLE
Push-to-talk functionality

### LINKED ISSUE #420: Add support for microphone input device selection
- [ ] Use an audio library, e.g. NAudio, to get a stream to pass into the System.Speech recognition engine
- [ ] Add an option to select the input device

## Nice to have:
- [ ] If the default is selected, automatically switch over input device when the default is changed

### PR DESCRIPTION
**Pull request by Vivelin**
I made a stab at push-to-talk functionality. It works on my machine™ but needs more work before this gets merged. I’ve created it as a draft but it’s technically release-ready — just lacking configurability (see below).

While Tracker constantly mishearing things when she’s not even being talked to can lead to funny situations, I want to offer this option. Especially because this way, input device selection should be possible since push-to-talk will result in a finite length Stream, which was the original limitation with #420.

However, I’ve mostly been working on stuff during my holidays, and I need a break too, so I was hoping @MattEqualsCoder or @CPColin could finish this up.

- [x] Separate push-to-talk implementation of SpeechRecognitionServiceBase
- [x] Detect key presses
- [x] Record microphone audio
- [x] Run recognition on recorded audio
- [x] Make push-to-talk hotkey configurable (currently hardcoded to left Ctrl)
- [x] Make input device configurable (currently hardcoded to default device)
    - Either something in the Options window or a right-click menu on the existing microphone status bar icon?
    - I run initialization during each key press event, so switching devices shouldn't require anything more beyond passing it to the `MicrophoneService` somehow, and leaving it at the default should also be fine.
- [x] Make speech recognition mode configurable: always-on / push-to-talk / none
- [ ] Configure/detect/negotiate the recording wave format
    - Currently I default to the default WaveFormat constructor which is 44.1KHz 16 bit and probably works for most people, but there might be some exotic setups where that isn’t supported or isn’t sufficient. There should be some negotation between what the selected input device supports and what System.Speech supports.

Resolves #420

### CODE REVIEW

File: src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public static IServiceCollection AddTracker(this IServiceCollection services)
             services.AddScoped<PushToTalkSpeechRecognitionService>();
             services.AddScoped<NullSpeechRecognitionService>();
             services.AddTransient<IMicrophoneService, MicrophoneService>();
-            services.AddSingleton<IGlobalHook, TaskPoolGlobalHook>();
+            services.AddScoped<IGlobalHook, TaskPoolGlobalHook>();
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs:50]: Why did you change this to scoped? The [library documentation](https://sharphook.tolik.io/v5.2.0/articles/hooks.html) specifically says to only have one instance active at a time in the application.

**Review by Vivelin — COMMENTED** [review]
I think it might be better to use the device IDs instead of looking up the devices by name. A dictionary that maps name->id should be sufficient, I guess?

File: src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs (hunk 1 above)
Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs:50]: Well, with the way this is setup and the fact you can only have a single tracker active, so it should only have one instance available at a time.

But, the issue was that it was crashing when closing tracker and re-opening because closing the prior tracker disposed of the hook. When I removed the dispose, it seemed to hang on closing tracker.

Maybe the dispose can be handled on closing the app, perhaps?

Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs:50]: Pushed in a fix for the id, but after messing with it a bit, I'm really just not familiar enough with this library you're using quickly discern best practice here. You have it trying to rerun the hook each time tracker is loaded, and if it's on a singleton that messes up on the second run even if I try to wait until app close to dispose of them.

I suppose I could have the hook and hook runner task initialized on the app launch, then the PushToTalkSpeechRecognitionService just binds/unbinds the events?

Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3.Tracking/TrackerServiceCollectionExtensions.cs]: Went ahead and did that. The hook and hook task are now initialized in the app and disposed of on exit. Tested, and I'm able to close tracker and reopen with no issues.

File: src/Randomizer.SMZ3.Tracking/Services/MicrophoneService.cs [hunk 2]
```diff
@@ -87,36 +87,36 @@ public void Dispose()
...
+        if (string.IsNullOrEmpty(id) || "Default" == id)
         {
             return WasapiCapture.GetDefaultCaptureDevice();
         }
 
         var enumerator = new MMDeviceEnumerator();
         return enumerator
             .EnumerateAudioEndPoints(DataFlow.Capture, DeviceState.Active)
-            .FirstOrDefault(wasapi => wasapi.DeviceFriendlyName == deviceName);
+            .FirstOrDefault(wasapi => wasapi.ID == id);
```
Comment (Vivelin) [inline src/Randomizer.SMZ3.Tracking/Services/MicrophoneService.cs:112]: There should still be a fallback to the default here in case the device gets disconnected/removed between setting it in Options and actually starting Tracker. Maybe with an error, logged and ideally shown to the user if possible.

**Comment by Vivelin** [comment]
Also, since I started the PR I can't approve it

File: src/Randomizer.SMZ3.Tracking/Services/MicrophoneService.cs (hunk 2 above)
Comment (MattEqualsCoder) [reply src/Randomizer.SMZ3.Tracking/Services/MicrophoneService.cs:112]: I already covered cases of the particular audio device not being found and no audio devices being found in GetInputDevice on lines 119 and 123. Though, I think a device can be returned but its status is marked as not active anymore, in the case of being disabled. I suppose I can try to account for that scenario as well and add logging around all this.

I'll see what I can do about displaying something to the user.

**Review by CPColin — APPROVED** [review]
Ship it!
